loadbalancer: Optimize ServiceName and L3n4Addr key construction

joamaki opened the pull request:
* Use weak.Pointer for container/cache to avoid ending up with many cache arrays when there are lots of cores. Since we're doing a lot of string dedup'ing already this should increase cache hit rates and reduce memory usage overall in the agent. Since we'll only hold onto one array, bump the cache size to 1024 entries to improve hit rate (8*1024=~8k per cache, and we have Strings, StringMap, ServiceCache and L3n4AddrBytes caches, so static allocation of ~32k for them all).
* Use a single string for ServiceName allowing direct casting into the StateDB key
* Use a cache for L3n4Addr Bytes() to greatly reduce constructing the StateDB from scratch
* Remove some old unused methods from `loadbalancer.go`
* Remove `k8s.ServiceID` and use `loadbalancer.ServiceName` instead in `redirectpolicy` and `k8s.Endpoints`

Before:
```
pkg: github.com/cilium/cilium/pkg/loadbalancer/writer
cpu: 13th Gen Intel(R) Core(TM) i9-13950HX
Benchmark_UpsertServiceAndFrontends_100-32          2673            424350 ns/op            235655 objects/sec    456719 B/op       6286 allocs/op
Avg: Allocated 685859kB in total, 2441351 objects / 166553kB still reachable (per service:  48 objs, 14046B alloc,  3411B in-use)
Avg: Reconciled 50000 objects in 846.8988ms  (16.937µs  per service /  59042 services per second)
```

After:
```
Benchmark_UpsertServiceAndFrontends_100-32          2842            408752 ns/op            244647 objects/sec    428990 B/op       5202 allocs/op
Avg: Allocated 646749kB in total, 2381278 objects / 162073kB still reachable (per service:  47 objs, 13245B alloc,  3319B in-use)
Avg: Reconciled 50000 objects in 831.240522ms (16.624µs  per service /  60154 services per second)

```

`Benchmark_UpsertServiceAndFrontends` allocates ~15% less, ~4% faster.
~6% less allocations and ~4% less resident memory use in `pkg/loadbalancer/benchmark`.

joamaki committed:
loadbalancer: Optimize ServiceName

Use a single string to hold the cluster/namespace/name along with
positions into it. This allows implementing String() and conversion
to index.Key without any allocations.

Change the ServiceName JSON serialization to just return the string,
the same as with YAML. This is only used for debugging purposes so
this is safe to do.

Result of this is that we can reduce allocations and copies when
computing the ServiceName StateDB keys.

Before:
pkg: github.com/cilium/cilium/pkg/loadbalancer
BenchmarkServiceNameKey-32   16476112       74.32 ns/op           32 B/op          2 allocs/op

pkg: github.com/cilium/cilium/pkg/loadbalancer/writer
Benchmark_UpsertServiceAndFrontends_100
Benchmark_UpsertServiceAndFrontends_100-32          1676            675638 ns/op            148008 objects/sec    456782 B/op       6286 allocs/op
Benchmark_UpsertServiceAndFrontends_1
Benchmark_UpsertServiceAndFrontends_1-32           57444             28206 ns/op             35453 objects/sec     33322 B/op        137 allocs/op
BenchmarkInsertBackend
BenchmarkInsertBackend-32                           1742            854584 ns/op            117016 objects/sec    617888 B/op       9108 allocs/op
BenchmarkReplaceBackend
BenchmarkReplaceBackend-32                        227928              5123 ns/op            195208 objects/sec      4834 B/op        100 allocs/op
BenchmarkReplaceService
BenchmarkReplaceService-32                       1265302               944.4 ns/op         1058866 objects/sec       536 B/op         12 allocs/op

After:
pkg: github.com/cilium/cilium/pkg/loadbalancer
BenchmarkNewServiceName-32   25104672       47.70 ns/op            0 B/op          0 allocs/op
BenchmarkServiceNameKey-32   25291726       47.49 ns/op            0 B/op          0 allocs/op

pkg: github.com/cilium/cilium/pkg/loadbalancer/writer
Benchmark_UpsertServiceAndFrontends_100
Benchmark_UpsertServiceAndFrontends_100-32          2022            654901 ns/op            152695 objects/sec    440281 B/op       5533 allocs/op
Benchmark_UpsertServiceAndFrontends_1
Benchmark_UpsertServiceAndFrontends_1-32           39508             29330 ns/op             34095 objects/sec     33389 B/op        129 allocs/op
BenchmarkInsertBackend
BenchmarkInsertBackend-32                           1250            843193 ns/op            118597 objects/sec    573891 B/op       8208 allocs/op
BenchmarkReplaceBackend
BenchmarkReplaceBackend-32                        241011              4805 ns/op            208119 objects/sec      4368 B/op         89 allocs/op
BenchmarkReplaceService
BenchmarkReplaceService-32                       1465656               855.1 ns/op         1169401 objects/sec       488 B/op         11 allocs/op

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Add benchmark for ServiceName StateDB key construction

Add a baseline benchmark to compare against in the next commit.

BenchmarkServiceNameKey-32   16476112                74.32 ns/op           32 B/op          2 allocs/op

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
container/cache: Add Cache.GetOrPut

Add a variant that looks up from the cache and if not found calls the
given function to "finalize" the object to be put into the cache. This
allows storing an object with some pre-computed fields that nevertheless
are not needed to look up from the cache.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Cache for L3n4Addr keys

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
container/cache: Use a weak pointer for cache array

The problem with sync.Pool is that the likelyhood of allocating new cache arrays goes
up quickly the more CPU cores we have. Ideally we would just have one per cache.

To fix this, use a weak.Pointer to hold onto a single copy of the cache.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Remove unused methods

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink commented on pkg/container/cache/cache.go:
```diff
@@ -17,19 +19,15 @@ func New[T any](hashfn func(T) uint64, skipfn func(x T) bool, eqfn func(a, b T)
-	// allows the Go runtime to drop the caches.
-	pool sync.Pool
+	mu  lock.Mutex
+	arr weak.Pointer[[cacheSize]T]
```
I wonder if `[cacheSize]weak.Pointer[T]` might not work better. Currently, you have an array of 512 values, the only reference to this array is the weak pointer so I expect the GC to often collect this array and you having to re-create it.

If you instead were to have an array of 512 weak pointers, then the GC should only prune the underlying entry if there exist no more strong references (the cache is the only reference).

That might increase hit rate. Just a thought

joamaki replied:
Hmm the array itself is the big allocation here and the elements themselves are usually small (e.g. a string is 16 bytes). Also doing `[cacheSize]weak.Pointer[T]` would mean every `T` is heap allocated so likely uses much more memory overall. If we'd be caching very large objects I think it'd make sense that way around.

joamaki replied:
With `weak.Pointer[[cachesize]T]`:
```
Benchmark_UpsertServiceAndFrontends_100-32          1974            570747 ns/op            175209 objects/sec    429561 B/op       5241 allocs/op
Min: Allocated 632178kB in total, 1919839 objects / 115710kB still reachable (per service:  38 objs, 12947B alloc,  2369B in-use)
Min: Reconciled 50000 objects in 795.845219ms (15.916µs  per service /  62830 services per second)
```
With `[cacheSize]weak.Pointer[T]`:
```
Benchmark_UpsertServiceAndFrontends_100-32          1952            567585 ns/op            176185 objects/sec    432226 B/op       5336 allocs/op
Min: Allocated 652386kB in total, 1720151 objects / 113297kB still reachable (per service:  34 objs, 13360B alloc,  2320B in-use)
Min: Reconciled 50000 objects in 910.93451ms (18.218µs  per service /  54891 services per second)
```

Not a huge difference, but it does allocate a bit more. Potentially holding onto less memory after GCing for some reason. It is noticeably slower (at least from the 3-4 runs I did) so I think it makes sense to go with `weak.Pointer[[cacheSize]T]`.

dylandreimerink replied:
> So I think it makes sense to go with `weak.Pointer[[cacheSize]T]`.

Lets. Thanks for testing out anyway.

joamaki committed:
k8s,loadbalancer: Remove k8s.ServiceID, use loadbalancer.ServiceName

Remove k8s.ServiceID and use the more optimized loadbalancer.ServiceName
instead.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
